File: clients/ios/Lextures/Core/LMS/IntegrationsAdminLogic.swift
import Foundation

/// Integrations & provisioning admin helpers (M14.8) — LTI, SCIM, cloud, LRS, OER.
enum IntegrationsAdminLogic {
    static let rbacManagePermission = "global:app:rbac:manage"

    enum Section: String, CaseIterable, Identifiable {
        case lti
        case scim
        case cloud
        case lrs
        case oer

        var id: String { rawValue }

        var titleKey: String.LocalizationValue {
            switch self {
            case .lti: "mobile.admin.integrations.lti.title"
            case .scim: "mobile.admin.integrations.scim.title"
            case .cloud: "mobile.admin.integrations.cloud.title"
            case .lrs: "mobile.admin.integrations.lrs.title"
            case .oer: "mobile.admin.integrations.oer.title"
            }
        }

        var subtitleKey: String.LocalizationValue {
            switch self {
            case .lti: "mobile.admin.integrations.lti.entry.subtitle"
            case .scim: "mobile.admin.integrations.scim.entry.subtitle"
            case .cloud: "mobile.admin.integrations.cloud.entry.subtitle"
            case .lrs: "mobile.admin.integrations.lrs.entry.subtitle"
            case .oer: "mobile.admin.integrations.oer.entry.subtitle"
            }
        }

        var systemImage: String {
            switch self {
            case .lti: "puzzlepiece.extension"
            case .scim: "person.2.badge.gearshape"
            case .cloud: "cloud"
            case .lrs: "arrow.triangle.branch"
            case .oer: "books.vertical"
            }
        }

        var webPath: String {
            switch self {
            case .lti: "/settings/lti-tools"
            case .scim: "/settings/scim-provisioning"
            case .cloud: "/settings/cloud-providers"
            case .lrs: "/settings/lrs-integrations"
            case .oer: "/settings/oer-providers"
            }
        }
    }

    static func adminSettingsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileAdminSettings || features.ffMobileAdminConsole
    }

    static func canManage(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission)
    }

    static func shouldShowEntry(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        !features.ffMobileAdminConsole && features.ffMobileAdminSettings && canManage(permissions: permissions)
    }

    static func canView(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        adminSettingsEnabled(features) && canManage(permissions: permissions)
    }

    /// LTI and cloud always available when admin entry is shown; SCIM/LRS/OER need extra flags.
    static func isSectionVisible(
        _ section: Section,
        features: MobilePlatformFeatures,
        scimEnabled: Bool
    ) -> Bool {
        switch section {
        case .lti, .cloud:
            return true
        case .scim:
            return scimEnabled
        case .lrs:
            return features.xapiEmissionEnabled
        case .oer:
            return features.oerLibraryEnabled
        }
    }

    static func visibleSections(
        features: MobilePlatformFeatures,
        scimEnabled: Bool
    ) -> [Section] {
        Section.allCases.filter { isSectionVisible($0, features: features, scimEnabled: scimEnabled) }
    }

    static func cloudProviderLabel(_ provider: String) -> String {
        switch provider {
        case "google_drive": return L.text("mobile.admin.integrations.cloud.googleDrive")
        case "onedrive": return L.text("mobile.admin.integrations.cloud.onedrive")
        case "dropbox": return L.text("mobile.admin.integrations.cloud.dropbox")
        default: return provider
        }
    }

    static func oerProviderLabel(_ provider: String) -> String {
        switch provider {
        case "oer_commons": return L.text("mobile.admin.integrations.oer.oerCommons")
        case "merlot": return L.text("mobile.admin.integrations.oer.merlot")
        case "openstax": return L.text("mobile.admin.integrations.oer.openstax")
        default: return provider
        }
    }

    static func activeTokenCount(_ tokens: [ScimTokenRow]) -> Int {
        tokens.filter { $0.revokedAt == nil || $0.revokedAt?.isEmpty == true }.count
    }

    static func lastEventAt(_ events: [ScimEventRow]) -> String? {
        events.first?.createdAt
    }

    static func formatTimestamp(_ raw: String?) -> String {
        let formatted = DateFormatting.formatDateTime(raw)
        return formatted.isEmpty ? L.text("mobile.emDash") : formatted
    }

    static func ltiActiveCount(platforms: [LtiParentPlatform], tools: [LtiExternalTool]) -> Int {
        platforms.filter(\.active).count + tools.filter(\.active).count
    }

    static func enabledCount(_ items: [(Bool)]) -> Int {
        items.filter { $0 }.count
    }

    static func enabledProviderCount(_ providers: [CloudProviderStatus]) -> Int {
        providers.filter(\.enabled).count
    }

    static func enabledLrsCount(_ endpoints: [LrsEndpointStatus]) -> Int {
        endpoints.filter(\.enabled).count
    }

    static func enabledOerCount(_ providers: [OerProviderStatus]) -> Int {
        providers.filter(\.enabled).count
    }

    /// Ensures mobile never treats secret-looking fields as present on status models.
    static func cloudStatusExcludesSecrets(_ jsonKeys: Set<String>) -> Bool {
        let forbidden = ["clientId", "apiKey", "appKey", "client_id", "api_key", "app_key"]
        return forbidden.allSatisfy { !jsonKeys.contains($0) }
    }

    static func applyingLtiPlatformActive(
        _ platforms: [LtiParentPlatform],
        id: String,
        active: Bool
    ) -> [LtiParentPlatform] {
        platforms.map { row in
            guard row.id == id else { return row }
            var copy = row
            copy.active = active
            return copy
        }
    }

    static func applyingLtiToolActive(
        _ tools: [LtiExternalTool],
        id: String,
        active: Bool
    ) -> [LtiExternalTool] {
        tools.map { row in
            guard row.id == id else { return row }
            var copy = row
            copy.active = active
            return copy
        }
    }

    static func applyingCloudEnabled(
        _ providers: [CloudProviderStatus],
        provider: String,
        enabled: Bool
    ) -> [CloudProviderStatus] {
        providers.map { row in
            guard row.provider == provider else { return row }
            var copy = row
            copy.enabled = enabled
            return copy
        }
    }

    static func applyingLrsEnabled(
        _ endpoints: [LrsEndpointStatus],
        id: String,
        enabled: Bool
    ) -> [LrsEndpointStatus] {
        endpoints.map { row in
            guard row.id == id else { return row }
            var copy = row
            copy.enabled = enabled
            return copy
        }
    }

    static func applyingOerEnabled(
        _ providers: [OerProviderStatus],
        provider: String,
        enabled: Bool
    ) -> [OerProviderStatus] {
        providers.map { row in
            guard row.provider == provider else { return row }
            var copy = row
            copy.enabled = enabled
            return copy
        }
    }

    static func userFacingError(_ error: Error, fallbackKey: String.LocalizationValue) -> String {
        if let api = error as? APIError, case let .httpStatus(_, message) = api, let message, !message.isEmpty {
            return message
        }
        return L.text(fallbackKey)
    }

    static func webHubPath() -> String { "/settings/lti-tools" }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/IntegrationsAdminLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.network.ApiError

/** Integrations & provisioning admin helpers (M14.8) — LTI, SCIM, cloud, LRS, OER. */
object IntegrationsAdminLogic {
    const val RBAC_MANAGE_PERMISSION = "global:app:rbac:manage"

    enum class Section {
        LTI,
        SCIM,
        CLOUD,
        LRS,
        OER,
        ;

        val webPath: String
            get() = when (this) {
                LTI -> "/settings/lti-tools"
                SCIM -> "/settings/scim-provisioning"
                CLOUD -> "/settings/cloud-providers"
                LRS -> "/settings/lrs-integrations"
                OER -> "/settings/oer-providers"
            }
    }

    fun adminSettingsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffMobileAdminSettings || features.ffMobileAdminConsole

    fun canManage(permissions: List<String>): Boolean =
        RBAC_MANAGE_PERMISSION in permissions

    fun shouldShowEntry(features: MobilePlatformFeatures, permissions: List<String>): Boolean =
        !features.ffMobileAdminConsole && features.ffMobileAdminSettings && canManage(permissions)

    fun canView(features: MobilePlatformFeatures, permissions: List<String>): Boolean =
        adminSettingsEnabled(features) && canManage(permissions)

    fun isSectionVisible(
        section: Section,
        features: MobilePlatformFeatures,
        scimEnabled: Boolean,
    ): Boolean = when (section) {
        Section.LTI, Section.CLOUD -> true
        Section.SCIM -> scimEnabled
        Section.LRS -> features.xapiEmissionEnabled
        Section.OER -> features.oerLibraryEnabled
    }

    fun visibleSections(
        features: MobilePlatformFeatures,
        scimEnabled: Boolean,
    ): List<Section> = Section.entries.filter { isSectionVisible(it, features, scimEnabled) }

    fun cloudProviderLabelKey(provider: String): String = when (provider) {
        "google_drive" -> "mobile_admin_integrations_cloud_googleDrive"
        "onedrive" -> "mobile_admin_integrations_cloud_onedrive"
        "dropbox" -> "mobile_admin_integrations_cloud_dropbox"
        else -> provider
    }

    fun oerProviderLabelKey(provider: String): String = when (provider) {
        "oer_commons" -> "mobile_admin_integrations_oer_oerCommons"
        "merlot" -> "mobile_admin_integrations_oer_merlot"
        "openstax" -> "mobile_admin_integrations_oer_openstax"
        else -> provider
    }

    fun activeTokenCount(tokens: List<ScimTokenRow>): Int =
        tokens.count { it.revokedAt.isNullOrEmpty() }

    fun lastEventAt(events: List<ScimEventRow>): String? =
        events.firstOrNull()?.createdAt

    fun ltiActiveCount(platforms: List<LtiParentPlatform>, tools: List<LtiExternalTool>): Int =
        platforms.count { it.active } + tools.count { it.active }

    fun enabledProviderCount(providers: List<CloudProviderStatus>): Int =
        providers.count { it.enabled }

    fun enabledLrsCount(endpoints: List<LrsEndpointStatus>): Int =
        endpoints.count { it.enabled }

    fun enabledOerCount(providers: List<OerProviderStatus>): Int =
        providers.count { it.enabled }

    /** Ensures mobile never models secret fields on status payloads. */
    fun cloudStatusExcludesSecrets(jsonKeys: Set<String>): Boolean {
        val forbidden = setOf("clientId", "apiKey", "appKey", "client_id", "api_key", "app_key")
        return forbidden.none { it in jsonKeys }
    }

    fun applyingLtiPlatformActive(
        platforms: List<LtiParentPlatform>,
        id: String,
        active: Boolean,
    ): List<LtiParentPlatform> = platforms.map {
        if (it.id == id) it.copy(active = active) else it
    }

    fun applyingLtiToolActive(
        tools: List<LtiExternalTool>,
        id: String,
        active: Boolean,
    ): List<LtiExternalTool> = tools.map {
        if (it.id == id) it.copy(active = active) else it
    }

    fun applyingCloudEnabled(
        providers: List<CloudProviderStatus>,
        provider: String,
        enabled: Boolean,
    ): List<CloudProviderStatus> = providers.map {
        if (it.provider == provider) it.copy(enabled = enabled) else it
    }

    fun applyingLrsEnabled(
        endpoints: List<LrsEndpointStatus>,
        id: String,
        enabled: Boolean,
    ): List<LrsEndpointStatus> = endpoints.map {
        if (it.id == id) it.copy(enabled = enabled) else it
    }

    fun applyingOerEnabled(
        providers: List<OerProviderStatus>,
        provider: String,
        enabled: Boolean,
    ): List<OerProviderStatus> = providers.map {
        if (it.provider == provider) it.copy(enabled = enabled) else it
    }

    fun userFacingError(error: Throwable, fallback: String): String {
        val api = error as? ApiError.HttpStatus
        val message = api?.message
        return if (!message.isNullOrBlank()) message else fallback
    }

    fun webHubPath(): String = "/settings/lti-tools"
}
